cpp/include/raft/spatial/knn/detail/ivf_flat_search.cuh

```diff
@@ -0,0 +1,1370 @@
+                                                              uint8_t* query_shared,
+                                                              const int loadDim)
+{
+  constexpr int veclen = 4;  // 16 uint8_t
+  uint32_t queryReg[veclen];
+  const int loadIndex = loadDim * veclen;
+  ldg(queryReg, reinterpret_cast<uint32_t const*>(query) + loadIndex);
+  sts(reinterpret_cast<uint32_t*>(query_shared) + loadIndex, queryReg);
+}
+
+template <>
+__device__ __forceinline__ void queryLoadToShmem<int8_t, 8>(const int8_t* const& query,
+                                                            int8_t* query_shared,
+                                                            const int loadDim)
+{
+  constexpr int veclen = 2;  // 8 int8_t
+  int32_t queryReg[veclen];
+  const int loadIndex = loadDim * veclen;
+  ldg(queryReg, reinterpret_cast<int32_t const*>(query) + loadIndex);
+  sts(reinterpret_cast<int32_t*>(query_shared) + loadIndex, queryReg);
+}
+
+template <>
+__device__ __forceinline__ void queryLoadToShmem<int8_t, 16>(const int8_t* const& query,
+                                                             int8_t* query_shared,
+                                                             const int loadDim)
+{
+  constexpr int veclen = 4;  // 16 int8_t
+  int32_t queryReg[veclen];
+  const int loadIndex = loadDim * veclen;
+  ldg(queryReg, reinterpret_cast<int32_t const*>(query) + loadIndex);
+  sts(reinterpret_cast<int32_t*>(query_shared) + loadIndex, queryReg);
+}
+
+template <int kUnroll,
+          int wordsPerVectorBlockDim,
+          typename Lambda,
+          int Veclen,
+          typename T,
+          typename AccT>
+struct loadAndComputeDist {
+  Lambda compute_dist;
+  AccT& dist;
+
+  __device__ __forceinline__ loadAndComputeDist(AccT& dist, Lambda op)
+    : dist(dist), compute_dist(op)
+  {
+  }
+
+  template <typename IdxT>
+  __device__ __forceinline__ void runLoadShmemCompute(const T* const& data,
+                                                      const T* query_shared,
+                                                      IdxT loadIndex,
+                                                      IdxT baseShmemIndex,
+                                                      IdxT iShmemIndex)
+  {
+    T encV[kUnroll][Veclen];
+    T queryRegs[kUnroll][Veclen];
+    constexpr int stride  = kUnroll * Veclen;
+    const int shmemStride = baseShmemIndex + iShmemIndex * stride;
+#pragma unroll
+    for (int j = 0; j < kUnroll; ++j) {
+      ldg(encV[j], data + (loadIndex + j * wordsPerVectorBlockDim) * Veclen);
+      const int d = shmemStride + j * Veclen;
+      lds(queryRegs[j], &query_shared[d]);
+#pragma unroll
+      for (int k = 0; k < Veclen; ++k) {
+        compute_dist(dist, queryRegs[j][k], encV[j][k]);
+      }
+    }
+  }
+
+  template <typename IdxT>
+  __device__ __forceinline__ void runLoadShflAndCompute(const T*& data,
```

Comment:
// Computes distance between one query vector (per thread) and one vector from the index. 
// It runs on a set of coordinates defined by baseLoadIndex and kUnroll parameter.